src/wp-includes/html-api/class-wp-html-tag-processor.php

```diff
@@ -1039,13 +1040,40 @@ private function parse_next_tag() {
 					'-' === $html[ $at + 2 ] &&
 					'-' === $html[ $at + 3 ]
 				) {
-					$closer_at = strpos( $html, '-->', $at + 4 );
-					if ( false === $closer_at ) {
-						return false;
+					$closer_at = $at + 4;
+
+					/*
+					 * Abruptly-closed empty comments are a sequence of dashes followed by `>`.
+					 */
+					$span_of_dashes = strspn( $html, '-', $closer_at );
+					if ( '>' === $html[ $closer_at + $span_of_dashes ] ) {
+						$at = $closer_at + $span_of_dashes + 1;
+						continue;
 					}
 
-					$at = $closer_at + 3;
-					continue;
+					/*
+					 * Comments may be closed by either a --> or an invalid --!>.
+					 * The first occurrence closes the comment.
+					 *
+					 * See https://html.spec.whatwg.org/#parse-error-incorrectly-closed-comment
+					 */
+					$closer_at--; // Pre-increment inside condition avoids risk of infinite looping.
+					while ( ++$closer_at < strlen( $html ) ) {
+						$closer_at = strpos( $html, '--', $closer_at );
+						if ( false === $closer_at ) {
+							return false;
+						}
+
+						if ( '>' === $html[ $closer_at + 2 ] ) {
```

Comment:
Is it possible that `$closer_at + 2` may not exist as an offset?

Reply:
yes it was, and thanks for noticing! I've added new tests and fixes in ae59e1a and 926d491fe3 which validate this particular case of unchecked indices